### PR TITLE
`Communication`: Fix issue causing incorrect user roles and authorities to be displayed

### PR DESCRIPTION
**Pull request by egekurt123**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.

### Description

This pull request addresses an issue in the Communication module where incorrect user roles and authorities were being displayed.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with communication enabled

1. Log in to Artemis
2. Navigate to Communication
3. Write a message with an instructor
4. Reply within a thread with a student role
5. Write another message with a student role
6. Reply with a instructor role
7. Make changes (react / pin / reply) to the first message
8. Change between threads
9. Make sure user role gets displayed correctly

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved handling of post data to resolve issues with WebSocket updates, ensuring accurate display of post authors, author roles, and answers based on cached data.

### CODE REVIEW

File: src/main/webapp/app/shared/metis/metis.service.ts [hunk 1]
```diff
@@ -597,6 +596,16 @@ export class MetisService implements OnDestroy {
             case MetisPostAction.UPDATE:
                 const indexToUpdate = this.cachedPosts.findIndex((post) => post.id === postDTO.post.id);
                 if (indexToUpdate > -1) {
+                    // WebSocket does not currently update the author and authorRole of posts correctly, so this is implemented as a workaround
+                    postDTO.post.author = this.cachedPosts[indexToUpdate].author;
```
Comment (anian03) [inline src/main/webapp/app/shared/metis/metis.service.ts:600]: Are you sure this is necessary? In my testing, the author attribute returned by the websocket was always there. Just authorRole was missing

**Review by anian03 — COMMENTED** [review]
Tested, everything seems to work. Just added one small question regarding overwriting the author of posts

**Review by EneaGore — COMMENTED** [review]
Tested on TS4. It works fine for all scenarios in the testing steps. But when editing a message as instructor/moderator it needs a refresh to show correctly from the person doing the editing, other users see it correctly without refreshing.

**Review by edkaya — COMMENTED** [review]
Tested on ts4, It works according to the test description. But I observed the following behaviour. To reproduce, you can follow this: 
- Write a message with an instructor
- Reply within a thread with a student role
- Write another message with a student role
- Reply with a instructor role
- As an instructor, confirm the new replied message to resolve the thread (the thread that is opened by the student)
- Switch to the first thread and come back to the second thread again
- See that user roles are broken again 

But if you navigate to another channel and come back (or just reload the page), the user roles are now correct again. 
Screenshot 2024-07-13 at 17 26 47

File: src/main/webapp/app/shared/metis/metis.service.ts (hunk 1 above)
Comment (egekurt123) [reply src/main/webapp/app/shared/metis/metis.service.ts]: Yes you are right, only authorRole is missing. I removed the unnecessary line

**Review by egekurt123 — COMMENTED** [review]
@edkaya Thank you for testing. The issue occurred because the authorRole of the changed reply wasn't updated correctly. I've now implemented a workaround for this as well, as the WebSocket does not send this information. It should be working fine now.

File: src/main/webapp/app/shared/metis/metis.service.ts [hunk 2]
```diff
@@ -597,6 +597,14 @@ export class MetisService implements OnDestroy {
             case MetisPostAction.UPDATE:
                 const indexToUpdate = this.cachedPosts.findIndex((post) => post.id === postDTO.post.id);
                 if (indexToUpdate > -1) {
+                    // WebSocket does not currently update the author and authorRole of posts correctly, so this is implemented as a workaround
```
Comment (anian03) [inline src/main/webapp/app/shared/metis/metis.service.ts:600]:
```suggestion
                    // WebSocket does not currently update the authorRole of posts correctly, so this is implemented as a workaround
```

**Review by anian03 — APPROVED** [review]
Works for me :)

**Review by edkaya — APPROVED** [review]
Thank you for fixing the further bug. Tested on ts4 again, works as expected

**Review by konrad2002 — APPROVED** [review]
manually tested on ts4, works as expected

**Review by EneaGore — APPROVED** [review]
Retested on TS4. LGTM.

**Pull request merged**